azafea/event_processors/endless/metrics/machine.py

```diff
@@ -34,6 +34,12 @@ class Machine(Base):
     demo = Column(Boolean, server_default=expression.false())
     dualboot = Column(Boolean, server_default=expression.false())
     live = Column(Boolean, server_default=expression.false())
+    location_id = Column(Unicode, index=True)
+    location_city = Column(Unicode, index=True)
+    location_state = Column(Unicode, index=True)
+    location_street = Column(Unicode, index=True)
+    location_country = Column(Unicode, index=True)
+    location_facility = Column(Unicode, index=True)
```

Comment:
I think we should store the entire location dictionary here. There is no guarantee which fields it will or will not contain: the client will just blindly submit whatever somebody puts into the configuration file.

Given a JSONB column with the complete dictionary, maybe these columns could be [generated columns](https://www.postgresql.org/docs/12/ddl-generated-columns.html), if our postgresql is new enough for that? … ah no, we have 11.8 and this feature is new in 12. Ah well.

I think it is valuable to have the columns broken out in this way – it makes them much easier to access, whether in SQL (with more familiar syntax) or in Metabase (which doesn't have any special support for JSONB columns in its query builder UI).

I wonder whether these indexes should be [partial](https://www.postgresql.org/docs/8.0/indexes-partial.html) and only index non-empty values. We only have location information for 241 machines, out of 853 739 machines total.

Reply:
I can add a field storing the JSONB, and keep the new fields with a partial index.

Reply:
Done. The dictionary is always updated to the latest version received, and columns are updated for keys included in the received dictionary.